Hw7 api

Pashacabu opened the pull request:
Basic usage of OkHttp, Retrofit and TMDB API

List of now playing movies (1 page) is loading
Movie details are loaded for every selected movie.

More functions are to be added further.
![Screenshot_1610306344](https://user-images.githubusercontent.com/70441207/104133002-e60e6b00-5391-11eb-8b7a-8383d93d7a09.png)
![Screenshot_1610306355](https://user-images.githubusercontent.com/70441207/104133003-e7d82e80-5391-11eb-8f91-d30657cb4a7e.png)

Pashacabu committed:
Tried to implement MVVM pattern
Got the idea but not so sure about realisation
Data is loaded through MyViewModel every time onViewCreated is called, don't know how to fix it.

Pashacabu committed:
Check if data was already loaded added.

Pashacabu committed:
Check if moviesListData is empty fixed

Pashacabu committed:
Implemented base usage of OkHTTP3, Retrofit and TMDB API.

Loading "Now playing" list (1 page)
Loading movie details for every movie on list.

More usage of API functions (further pages loading, switching between now playing/popular/latest..., movie search, actor additional data) is yet to be done.

Pashacabu committed:
Implemented base usage of OkHTTP3, Retrofit and TMDB API.

Loading "Now playing" list (1 page)
Loading movie details for every movie on list.

More usage of API functions (further pages loading, switching between now playing/popular/latest..., movie search, actor additional data) is yet to be done.

PaulXIII commented on app/src/main/java/com/pashcabu/hw2/Cast.kt:
```diff
@@ -0,0 +1,46 @@
+	val actorPhoto: String? = null,
+//	val id: Int? = null,
+//	val adult: Boolean? = null,
+//	val order: Int? = null
```
Стоит удалить весь закоментированный код

Pashacabu replied:
Я пока его оставил на случай, если при добавлении ещё функций потребуются эти поля

PaulXIII commented on app/src/main/java/com/pashcabu/hw2/MovieDetailsResponse.kt:
```diff
@@ -0,0 +1,70 @@
+	@SerialName("adult")
+	val adult: Boolean? = null,
+//	val homepage: String? = null,
+//	val status: String? = null
```
Стоит удалить весь закоментированный код

Pashacabu replied:
Здесь тоже пока оставил на случай надобности

PaulXIII commented on app/src/main/java/com/pashcabu/hw2/MovieListResponse.kt:
```diff
@@ -0,0 +1,48 @@
+	val posterPath: String? = null,
+//	val backdropPath: String? = null,
+//	val releaseDate: String? = null,
+//	val popularity: Double? = null,
```
Стоит удалить весь закоментированный код

PaulXIII commented on app/src/main/java/com/pashcabu/hw2/MovieDetailsFragment.kt:
```diff
@@ -142,4 +151,7 @@ class MovieDetailsFragment : Fragment() {
+
+}
+
+const val imageBaseUrl = "https://image.tmdb.org/t/p/w780"
```
Константы лучше хранить в `companion object`

PaulXIII commented on app/src/main/java/com/pashcabu/hw2/NetworkModule.kt:
```diff
@@ -0,0 +1,61 @@
+    }
+}
+
+const val api_key = "690d3ea0f7ef1f69512be4c95fc7a886"
```
Константы лучше хранить в `companion object`

PaulXIII commented on app/src/main/java/com/pashcabu/hw2/recyclerAdapters/MovieDetailsAdapter.kt:
```diff
@@ -0,0 +1,65 @@
+    fun onActorSelected(actor: CastItem?)
+}
+
+const val baseImageUrl = "https://image.tmdb.org/t/p/w185"
```
Константы лучше хранить в `companion object`

PaulXIII commented on app/src/main/java/com/pashcabu/hw2/recyclerAdapters/MoviesListAdapter.kt:
```diff
@@ -0,0 +1,93 @@
+
+}
+
+const val imageBaseUrl = "https://image.tmdb.org/t/p/w185"
```
Константы лучше хранить в `companion object`

PaulXIII commented on app/src/main/java/com/pashcabu/hw2/MyViewModel.kt:
```diff
@@ -0,0 +1,75 @@
+                            NetworkModule().apiService.getGenres(api_key).genres?.associateBy({ it?.genreId },
+                                { it?.genreName })
+                        Log.d("Genres", "are loading")
+                        val listOfMovies = NetworkModule().apiService.getMoviesList(api_key, pageNumber)
```
`NetworkModule()` можно сделать глобальной переменной

PaulXIII commented on app/src/main/java/com/pashcabu/hw2/MyViewModel.kt:
```diff
@@ -0,0 +1,75 @@
+        if (mutableMovieDetailsData.value == null) {
+            viewModelScope.launch {
+                try {
+                    if (mutableMovieDetailsData.value==null){
```
А стоит проверять `mutableMovieDetailsData.value` дважды?

Pashacabu replied:
Здесь, конечно нет, это я зря)

Pashacabu committed:
Added few more features using API (in TabsLayout)

Pashacabu committed:
Structured code
Tabs for different movies lists
Load more movies when scrolled down
No internet connection crash fixed
Getting ready to implement DB for storing the lists and add movies to favourite

PaulXIII commented on app/src/main/java/com/pashcabu/hw2/views/adapters/NewMoviesListAdapter.kt:
```diff
@@ -0,0 +1,139 @@
+            if (totalItems != null) {
+                if (lastVisible == totalItems - 1) {
+                    Log.d("Adapter", "its time to load more")
+                    MyViewModel().showLoading()
```
А точно здесь нужно создавать новую `MyViewModel`?

PaulXIII commented on app/src/main/java/com/pashcabu/hw2/views/MoviesListFragment.kt:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    private fun addLoadMoreListener() {
+        val listener = MyScrollListener {
```
Похоже в `MyScrollListener` не хватает `callback`
